src/Reader.h

```diff
@@ -0,0 +1,36 @@
+#include <Rcpp.h>
+#include "Collector.h"
+#include "Source.h"
+#include "Progress.h"
+
+using namespace Rcpp;
+
+class Reader {
+  public:
+    Reader(SourcePtr source, TokenizerPtr tokenizer, std::vector<CollectorPtr> collectors,
+        CharacterVector colNames, LocaleInfo* locale, int progress = 10000);
+
+    RObject readToDataFrame(int lines = -1);
+
+    template <typename T>
+    T readToVector(int lines) {
+      read(lines);
+
+      return as<T>(collectors_[0]->vector());
```

Comment:
Maybe this should power `parse_vector_()` too so we should at least have `readToVectorWithWarnings()`?